datalad/tests/utils_cached_dataset.py

```diff
@@ -0,0 +1,233 @@
+    # things. It's about datasets used in the tests - they shouldn't change too
+    # frequently.
+    elif any('uptodate' not in c['operations']
+             for c in ds.repo.fetch('origin')):
+        rmtree(ds.path)
+        ds = Clone()(url, ds.pathobj)
+
+    if version:
+        # check whether version is available
+        assert ds.repo.commit_exists(version)
+    if keys and AnnexRepo.is_valid_repo(ds.path):
+        ds.repo.get(keys, key=True)
+
+    return ds
+
+
+@optional_args
+def cached_dataset(f, url=None, version=None, paths=None):
+    """Test decorator providing a clone of `url` from cache
+
+    If config datalad.tests.cache is not set, delivers a clone in a temporary
+    location of the original `url`. Otherwise that clone is in fact a clone of a
+    cached dataset (origin being the cache instead of `url`).
+    This allows to reduce time and network traffic when using a dataset in
+    different tests.
+
+    The clone will checkout `version` and get the content for `paths`.
+
+    Parameters
+    ----------
+    url: str
+        URL to the to be cloned dataset
+    version: str
+        committish to checkout in the clone
+    paths: str or list
+        annexed content to get
+
+    Returns
+    -------
+    Dataset
+        a clone of the dataset at `url` at a temporary location (cleaned up,
+        after decorated test is finished - see with_tempfile). If caching is
+        enabled, it's actually a clone of a clone, 'origin' being the clone in
+        cache rather than the original repo at `url`.
+    """
+    @better_wraps(f)
+    @with_tempfile
+    def newfunc(*arg, **kw):
+
+        if DATALAD_TESTS_CACHE:
+            # Note: We can't pass keys based on `paths` parameter to
+            # get_cached_dataset yet, since translation to keys depends on a
+            # worktree. We'll have the worktree of `version` only after cloning.
+            ds = get_cached_dataset(url, version=version)
+            clone_ds = Clone()(ds.pathobj, arg[-1])
+        else:
+            clone_ds = Clone()(url, arg[-1])
+        if version:
+            clone_ds.repo.checkout(version)
+        if paths and AnnexRepo.is_valid_repo(clone_ds.path):
+            # just assume ds is annex as well. Otherwise `Clone` wouldn't
+            # work correctly - we don't need to test its implementation here
+            if DATALAD_TESTS_CACHE:
+                # cache is enabled; we need to make sure it has the desired
+                # content, so clone_ds can get it from there. However, we got
+                # `paths` and potentially a `version` they refer to. We can't
+                # assume the same (or any) worktree in cache. Hence we need to
+                # translate to keys.
+                keys = clone_ds.repo.get_file_key(paths)
+                ds.repo.get(keys, key=True)
+                clone_ds.repo.fsck(remote='origin', fast=True)
+
+            clone_ds.get(ensure_list(paths))
```

Comment:
Doesn't `get` already handle the string case?

Reply:
True. That seems superfluous.